### ISSUE TITLE
The MULTISET JSON emulation doesn't work with DISTINCT in PostgreSQL

### ISSUE THREAD

**Issue by jippeholwerda** [jOOQ/jOOQ#13805]
After upgrading from 3.16 to 3.17.2 the following jOOQ code gives an error in Postgres 13:

```kotlin
context.selectDistinct(
            ZKN_ZAAK.ID.convertFrom { ZaakId(it!!) },
            multiset(
                permissieFields(context)
                    .from(ZKN_AUTORISATIE)
                    .join(STD_PERMISSIE).onKey()
                    .where(ZKN_ROL.ROLTYPE_ALIAS_ID.eq(ZKN_AUTORISATIE.ROLTYPE_ID))
            ).convertFrom { it.map(mapping(::Permissie)) },
        )
            .from(ZKN_ZAAK)
            .leftJoin(ZKN_ROL).on(ZKN_ZAAK.ID.eq(ZKN_ROL.ZAAK_ID).and(ZKN_ROL.MEDEWERKER_ID.eq(medewerkerId.value)))
            .where(
                ZKN_ZAAK.ID.`in`(
                    context.select(ZKN_ZAAK.ID).from(ZKN_ZAAK)
                        .join(ZKN_ZAAKOBJECT).onKey()
                        .join(REG_LOCATIE).onKey()
                        .join(REG_MBA).on(REG_LOCATIE.ID.eq(REG_MBA.LOCATIE_ID))
                        .where(REG_MBA.ID.eq(mbaId.value))
                )
            )
            .fetchGroups({ it.value1() }, { it.value2() })
            .mapValues { it.value.flatten().filterNotNull().toSet() }
```

The error given is: `[42883] ERROR: could not identify an equality operator for type json Position: 50`

Generated SQL:

```sql
select distinct "zkn_zaak"."id",
                (select coalesce(json_agg(json_build_array("v0", "v1", "v2", "v3", "v4")), json_build_array())
                 from (select "std_permissie"."id"              as "v0",
                              "std_permissie"."resource"        as "v1",
                              "std_permissie"."actie"           as "v2",
                              "zkn_autorisatie"."precondities"  as "v3",
                              "zkn_autorisatie"."postcondities" as "v4"
                       from "zkn_autorisatie"
                                join "std_permissie" on "zkn_autorisatie"."permissie_id" = "std_permissie"."id"
                       where "zkn_rol"."roltype_alias_id" = "zkn_autorisatie"."roltype_id") as "t")
from "zkn_zaak"
         left outer join "zkn_rol" on ("zkn_zaak"."id" = "zkn_rol"."zaak_id" and "zkn_rol"."medewerker_id" = 'd71c627a-03c3-4070-a2d2-68c077a1bef8')
where "zkn_zaak"."id" in (select "zkn_zaak"."id"
                          from "zkn_zaak"
                                   join "zkn_zaakobject" on "zkn_zaakobject"."zaak_id" = "zkn_zaak"."id"
                                   join "reg_locatie" on "zkn_zaakobject"."locatie_id" = "reg_locatie"."id"
                                   join "reg_mba" on "reg_locatie"."id" = "reg_mba"."locatie_id"
                          where "reg_mba"."id" = 'c3a1013e-3605-4d0f-9236-38958859f5b4')
```

Using jsonb instead of json solves the issue.

**Comment by jippeholwerda** [comment]
You're absolutely right. Changing the query solves the issue and is indeed the cleaner solution compared to using `DISTINCT`.

**Comment by lukaseder** [comment]
Looking into this now. There are a set of preconditions that will need to force the `MULTISET` emulation locally from `JSON` to `JSONB` in PostgreSQL's case. For example, even this doesn't work:

```sql
select '[1]'::json = '[1]'::json
```

Something like that could appear when comparing `MULTISET` values with `Field::eq`, which is also an edge case, but could happen. In all of these cases, switching back to `JSONB` emulations would help. But I wouldn't want to make `JSONB` the default in general. The performance improvements are too significant.

Since people might choose `JSON` independently of the default, I think these edge cases need to be addressed anyway, also because maybe, some other RDBMS may require handling these.

**Comment by lukaseder** [comment]
Related problem, comparing XMLs isn't implemented either, in PostgreSQL (and others): https://github.com/jOOQ/jOOQ/issues/12074

**Comment by lukaseder** [comment]
In the `XML` case, one option would be to cast the document to `TEXT`, but that would make it harder to embed in other `XML` documents as we'd have to remember to cast it back from `TEXT` to `XML`.

With `JSON`/`JSONB`, we can "hope" that all `JSON_XYZ()` functions are overloaded to accept also `JSONB_XYZ()` and vice versa. So far, I haven't run into limitations, but quite probably, limitations are lurking somewhere. So, in the worst case, `JSON` / `JSONB` conversions would have to be reverted explicitly as well, just as with `XML` / `TEXT` conversions.

I'm not sure I like where this is going. Will reflect on this a bit longer.

**Comment by lukaseder** [comment]
While MySQL and SQL Server can compare JSON values, SQL Server cannot compare XMLs either, and Oracle cannot compare both XMLs and JSON values.

I guess the fact that this works for PostgreSQL's `JSONB` can be seen as incidental. I'm not sure if jOOQ should invest heavily into working around these limitations.

Obviously, the regression when upgrading to jOOQ 3.17 is annoying. The workaround is to revert #13650 by manually specifying the `NestedCollectionEmulation.JSONB` emulation. The overall performance benefit of using `NestedCollectionEmulation.JSON` as the default is still worth it, in my opinion. I'll leave this issue for now, hoping it will attract more feedback from the community.

**Comment by lukaseder** [comment]
After some additional thought, I've decided to revert the change from `JSONB` to `JSON` as the default emulation (https://github.com/jOOQ/jOOQ/issues/13650). Starting with jOOQ 3.18, `JSONB` will be used again, as default (https://github.com/jOOQ/jOOQ/issues/14641)

This issue here stays open, because we obviously want to make these things work also for `JSON` (or `XML`) emulations, but fixing the out of the box experience is simpler with #14641
